src/csrc/alphafold/fused_gating_attention_fwd_tmpl.h

```diff
@@ -152,154 +152,231 @@ float alpha = (1.0 / sqrt(key_dim));
+  // }
 
-{
-  RECORD_SCOPE(alpha_ac_gemm, {q, k, bias});
   {
-    RECORD_FUNCTION("parallel_for", std::vector<c10::IValue>());
-#pragma omp parallel for collapse(3)
-    for (int i = 0; i < B_t; i++) {
-      for (int n = 0; n < N_t; n++) {
-        for (int j1 = 0; j1 < S_t; j1 += A_BLOCKSIZE) {
-          T tmp_qv[A_BLOCKSIZE * H_t];
-          T tmp_logits[A_BLOCKSIZE][S_t];
-          // T tmp_o1[A_BLOCKSIZE * H_t];
-          // T tmp_o2[A_BLOCKSIZE * H_t];
-          // T tmp_S[A_BLOCKSIZE * Ak_BLOCKSIZE];
-          // float omax[A_BLOCKSIZE], osum[A_BLOCKSIZE], cmax[A_BLOCKSIZE],
-          // csum[A_BLOCKSIZE];
-
-          a_cpy_tpp(&q_a[i][j1][n][0], &tmp_qv[0]);
-          for (int j2 = 0; j2 < S_t; j2 += A_BLOCKSIZE) {
-            a_brgemm_tpp(
-                &tmp_qv[0], &k_a[i][n * H_t * S_t + j2], &tmp_logits[0][j2], 1);
-
-            //   a_addbias_tpp(&bias_a[i][j2], &tmp_logits[0][j2]);
-            //   if (flag)
-            //     a_add_nbbias_tpp(
-            //         &nonbatched_bias_a[0][n][j1][j2],
-            //         &tmp_logits[0][j2],
-            //         &tmp_logits[0][j2]);
-
-            //   a_cpy4_tpp(&tmp_logits[0][j2], tmp_S);
-            //   float *pmax, *psum;
-            //   if (j2 == 0) {
-            //     pmax = omax;
-            //     psum = osum;
-            //   } else {
-            //     pmax = cmax;
-            //     psum = csum;
-            //   }
-
-            //   a_softmax2_tpp(1, tmp_S, tmp_S, pmax, psum);   // calculate
-            //   local softmax a_brgemm2_tpp(tmp_S, &v_a[i][j2][n][0], tmp_o1,
-            //   1);      // O = P*V if (j2 == 0) {
-            //     a_cpy3_tpp(tmp_o1, tmp_o2);
-            //   } else {
-            //     a_softmax_fixup(tmp_o1, tmp_o2, cmax, csum, omax, osum);
-            //   }
-          }
-          a_addbias2_tpp(&bias_a[i][0], &tmp_logits[0][0]);
-          if (flag)
-            a_add_nbbias2_tpp(
-                &nonbatched_bias_a[0][n][j1][0],
-                &tmp_logits[0][0],
-                &tmp_logits[0][0]);
-
-          if (S_t == Sp_t) {
-            a_softmax_tpp(1, &tmp_logits[0][0], &tmp_logits[0][0]);
-          } else {
-            a_add_sfmask_tpp(&sfmask_a[0][0], &tmp_logits[0][Sp_t]);
-            a_softmax_tpp(1, &tmp_logits[0][0], &tmp_logits[0][0]);
+    RECORD_SCOPE(alpha_ac_gemm, {q, k, bias});
+    {
+      RECORD_FUNCTION("parallel_for", std::vector<c10::IValue>());
+  #pragma omp parallel for collapse(3)
+      for (int i = 0; i < B_t; i++) {
+        for (int n = 0; n < N_t; n++) {
+          for (int j1 = 0; j1 < S_t; j1 += A_BLOCKSIZE) {
+            T tmp_qv[A_BLOCKSIZE * H_t];
+            T tmp_o1[A_BLOCKSIZE * H_t];
+            T tmp_o2[A_BLOCKSIZE * H_t];
+            T tmp_S[A_BLOCKSIZE * Ak_BLOCKSIZE];
+            float omax[A_BLOCKSIZE], osum[A_BLOCKSIZE], cmax[A_BLOCKSIZE],
+            csum[A_BLOCKSIZE];
+
+            a_cpy_tpp(&q_a[i][j1][n][0], &tmp_qv[0]);
```

Comment:
Similar comments as previous regarding copies and zeros...

Reply:
Removed the unnecessary copies and zeros in the latest commit